### PR TITLE
Params definition for PSE module with excluded addresses

### PR DESCRIPTION
**Pull request by TxCorpi0x**
# Description

This pull request introduces a new parameter management system for the PSE module, focusing on a more robust and flexible way to manage excluded addresses via governance. The changes replace the previous generic parameter update mechanism with a dedicated structure and message for updating excluded addresses, and implement full support for storing, querying, and updating these parameters. Comprehensive tests and documentation updates are included.

**Parameter Management and Governance**

- Introduced a new `Params` structure in `params.proto` to hold module parameters, specifically the `excluded_addresses` list, and updated all relevant proto files and documentation to use this structure.
- Replaced the generic `MsgUpdateParams` with `MsgUpdateExcludedAddresses`, a governance message specifically for modifying the list of excluded addresses, and updated all references, handlers, and documentation accordingly.

**Keeper and Service Implementation**

- Added full keeper support for parameter storage and manipulation, including `GetParams`, `SetParams`, and `UpdateExcludedAddresses` methods, and wired these into the module’s gRPC query and message servers. 

**Genesis and State Management**

- Updated genesis handling to initialize and export the new `Params` structure, ensuring excluded addresses are properly persisted across chain restarts.

**Testing and Validation**

- Added comprehensive unit tests for parameter get/set and excluded address update logic, including authority checks and edge cases.

**Error Handling**

- Introduced a new `ErrInvalidAuthority` error for improved governance security and feedback.

# Reviewers checklist:
- [ ] Try to write more meaningful comments with clear actions to be taken.
- [ ] Nit-picking should be unblocking. Focus on core issues.

# Authors checklist
- [ ] Provide a concise and meaningful description
- [ ] Review the code yourself first, before making the PR.
- [ ] Annotate your PR in places that require explanation.
- [ ] Think and try to split the PR to smaller PR if it is big.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tokenize-x/tx-chain/11)

### CODE REVIEW

**Review by ysv — CHANGES_REQUESTED** [review]
> @ysv reviewed 23 of 23 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tokenize-x/tx-chain/11)* status: all files reviewed, 2 unresolved discussions (waiting on @miladz68)

___
*[`proto/tx/pse/v1/params.proto` line 12 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-Ocutn-g4w7Olfbfbvql:-Ocutn-g4w7Olfbfbvqm:buikfv9) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/proto/tx/pse/v1/params.proto#L12)):*
> ```proto
> message Params {
>   // excluded_addresses is a list of addresses excluded from PSE distribution.
>   // This list includes module accounts and other addresses that should not receive PSE rewards.
> ```

good point about module accounts.
Should we exclude them by default (independently on this list) or add them to this list ?



_Code quote:_
```proto
  // This list includes module accounts and other addresses that should not receive PSE rewards.
```
___
*[`x/pse/keeper/params.go` line 34 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-OcuuKzd-ltq0XNBAQvA:-OcuuKzd-ltq0XNBAQvB:b-5svtog) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/x/pse/keeper/params.go#L34)):*
> ```Go
> 	ctx context.Context,
> 	authority string,
> 	addressesToAdd, addressesToRemove []string,
> ```

what if 

1. addressesToAdd contains duplicate ?
2. what if address is included in both ToAdd & ToRemove

**Review by TxCorpi0x — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tokenize-x/tx-chain/11)* status: all files reviewed, 2 unresolved discussions (waiting on @miladz68 and @ysv)

___
*[`proto/tx/pse/v1/params.proto` line 12 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-Ocutn-g4w7Olfbfbvql:-Ocynq2j0NQlt0nW4OSu:bfxs9jh) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/proto/tx/pse/v1/params.proto#L12)):*
<details><summary><i>Previously, ysv (Yaroslav Savchuk) wrote…</i></summary><blockquote>

good point about module accounts.
Should we exclude them by default (independently on this list) or add them to this list ?
</blockquote></details>

We have two ways: 
1. Append a list of addresses in the blockchain startup
2. Include the addresses in the list in the upgrade handlers
___
*[`x/pse/keeper/params.go` line 34 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-OcuuKzd-ltq0XNBAQvA:-OcynbW0B2K61RyQCH-k:bykzkvx) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/x/pse/keeper/params.go#L34)):*
<details><summary><i>Previously, ysv (Yaroslav Savchuk) wrote…</i></summary><blockquote>

what if 

1. addressesToAdd contains duplicate ?
2. what if address is included in both ToAdd & ToRemove
</blockquote></details>

Two layers of validation in ValidateBasic and Params validation prevents all of wrong data scenarios, there are also tests in types and keeper packages.

**Review by miladz68 — CHANGES_REQUESTED** [review]
> @miladz68 reviewed 23 of 23 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tokenize-x/tx-chain/11)* status: all files reviewed, 3 unresolved discussions (waiting on @ysv)

___
*[`proto/tx/pse/v1/params.proto` line 12 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-Ocutn-g4w7Olfbfbvql:-Od7f3kD5p3RRLVq1DeO:b-ocp61x) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/proto/tx/pse/v1/params.proto#L12)):*
<details><summary><i>Previously, TxCorpi0x wrote…</i></summary><blockquote>

We have two ways: 
1. Append a list of addresses in the blockchain startup
2. Include the addresses in the list in the upgrade handlers
</blockquote></details>

I don't think module accounts should be part of this param field. they should be dynamically appended on demand.
___
*[`x/pse/module.go` line 111 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-Od7iE75EBJHNy9ybPHX:-Od7iE75EBJHNy9ybPHY:b3bbcke) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/x/pse/module.go#L111)):*
> ```Go
> // InitGenesis performs genesis initialization for the module. It returns
> // no validator updates.
> func (am AppModule) InitGenesis(ctx sdk.Context, cdc codec.JSONCodec, data json.RawMessage) {
> ```

you should add genesis test.

**Review by TxCorpi0x — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tokenize-x/tx-chain/11)* status: 18 of 25 files reviewed, 6 unresolved discussions (waiting on @Copilot, @miladz68, and @ysv)

___
*[`proto/tx/pse/v1/params.proto` line 12 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-Ocutn-g4w7Olfbfbvql:-Od9OnCR56LCZq3_P3MI:b-4zp829) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/proto/tx/pse/v1/params.proto#L12)):*
<details><summary><i>Previously, miladz68 (milad) wrote…</i></summary><blockquote>

I don't think module accounts should be part of this param field. they should be dynamically appended on demand.
</blockquote></details>

Removed from the comment.
___
*[`x/pse/module.go` line 111 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-Od7iE75EBJHNy9ybPHX:-Od9OxTlDrUmHfiwS8E1:bb6un1j) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/x/pse/module.go#L111)):*
<details><summary><i>Previously, miladz68 (milad) wrote…</i></summary><blockquote>

you should add genesis test.
</blockquote></details>

Done

File: x/pse/keeper/params.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+	// Get current params
+	params, err := k.GetParams(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Remove addresses first
+	params.ExcludedAddresses = lo.Filter(params.ExcludedAddresses, func(addr string, _ int) bool {
+		return !slices.Contains(addressesToRemove, addr)
```
[the remark this answers is not in the document]
Comment (ysv) [reply x/pse/keeper/params.go]: good point

File: x/pse/keeper/params.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	})
+
+	// Add new addresses only if they don't already exist
+	toActuallyAdd := lo.Filter(addressesToAdd, func(addr string, _ int) bool {
+		return !slices.Contains(params.ExcludedAddresses, addr)
```
[the remark this answers is not in the document]
Comment (ysv) [reply x/pse/keeper/params.go]: good point

**Review by ysv — CHANGES_REQUESTED** [review]
> @ysv reviewed 3 of 7 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tokenize-x/tx-chain/11)* status: all files reviewed, 5 unresolved discussions (waiting on @miladz68 and @TxCorpi0x)

___
*[`proto/tx/pse/v1/params.proto` line 12 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-Ocutn-g4w7Olfbfbvql:-Od9SAJp4R4XDjXA9KrF:bm0k64u) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/proto/tx/pse/v1/params.proto#L12)):*
<details><summary><i>Previously, TxCorpi0x wrote…</i></summary><blockquote>

Removed from the comment.
</blockquote></details>

so does it mean that we exclude module accounts inside code by default independently on this list ?
@miladz68 do you want to add this details to a task ?

File: x/pse/keeper/params.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (TxCorpi0x) [reply x/pse/keeper/params.go]: Done

File: x/pse/keeper/params.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (TxCorpi0x) [reply x/pse/keeper/params.go]: Done

**Review by TxCorpi0x — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tokenize-x/tx-chain/11)* status: 24 of 25 files reviewed, 3 unresolved discussions (waiting on @miladz68 and @ysv)

**Review by miladz68 — APPROVED** [review]
> @miladz68 reviewed 3 of 7 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tokenize-x/tx-chain/11)* status: all files reviewed, 2 unresolved discussions (waiting on @TxCorpi0x and @ysv)

___
*[`proto/tx/pse/v1/params.proto` line 12 at r1](https://reviewable.io/reviews/tokenize-x/tx-chain/11#-Ocutn-g4w7Olfbfbvql:-OdCoVoW8fgT8YTbDT4n:b-os96wa) ([raw file](https://github.com/tokenize-x/tx-chain/blob/2affbdc7cf703ddca0512452123a7c681fa38a2a/proto/tx/pse/v1/params.proto#L12)):*
<details><summary><i>Previously, ysv (Yaroslav Savchuk) wrote…</i></summary><blockquote>

so does it mean that we exclude module accounts inside code by default independently on this list ?
@miladz68 do you want to add this details to a task ?
</blockquote></details>

I will add a task to double check it and make sure we do not forget it.

**Review by ysv — APPROVED** [review]
> @ysv reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tokenize-x/tx-chain/11)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @TxCorpi0x)
